railgun/internal/ctrlutils/ingress-status.go

```diff
@@ -0,0 +1,388 @@
+package ctrlutils
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/kong/deck/file"
+	"github.com/kong/kubernetes-ingress-controller/pkg/sendconfig"
+	kicclientset "github.com/kong/kubernetes-ingress-controller/railgun/pkg/clientset"
+	"github.com/prometheus/common/log"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	knative "knative.dev/networking/pkg/apis/networking/v1alpha1"
+	knativeversioned "knative.dev/networking/pkg/client/clientset/versioned"
+	knativeApis "knative.dev/pkg/apis"
+	"knative.dev/pkg/network"
+)
+
+// dedicated function that process ingress/customer resource status update after configuration is updated within kong.
+func PullConfigUpdate(kongConfig sendconfig.Kong, log logr.Logger, ctx context.Context, kubeConfig *rest.Config, stopCh <-chan struct{}) {
+	log.Info("Launching Ingress Status Update Thread.")
+	var wg sync.WaitGroup
```

Comment:
I'm not seeing that we are actually using this `sync.WaitGroup` to any effect, as [defined in the documentation](https://golang.org/src/sync/waitgroup.go) the point is to allow a caller to create several concurrent processes (goroutines in this case) and then `Wait()` for them to complete before proceeding.

When I first wrote this, we were never running `Wait()`, but as I was writing this we updated to waiting right after we fire off the goroutine. We don't appear to be using this waitgroup to any actual effect?

Reply:
@ccfishk this comment was resolved but without any updates to the comment thread. Can you help me to understand what the resolution is here please?

Reply:
We dont have static goroutine number since it is a dynamic and flow process. 
To avoid goroutiene leak, apply following pattern would be good enough.
wg.Add(1)
go func(&wg){wg.Done()}
wg.Wait()

Reply:
:thinking: applying that pattern would seem to use the waitgroup without a purpose, if you add to the waitgroup, but then wait for the waitgroup, it's equivalent to just running the function synchronously, right?

Reply:
No. here is goroutine, is a thread, asynchronous mechanism for concurrency. We support as many ingress as possible, cant use synchronous function.

Reply:
> No. here is goroutine, is a thread, asynchronous mechanism for concurrency. We support as many ingress as possible, cant use synchronous function.

In its current state this **is a synchronous function**, so if you're saying we can't use a synchronous function here, then we have to make some kind of change because the call to wait _will block_ until the call to `UpdateIngress` completes.

Reply:
Each goroutine is a thread, which is asynchronous function, synchronous function would block the update and generate performance issue. 
Also, communicated this to Shane offline.